### PR TITLE
turn off hybrid evaluation by masking the function. #3255

### LINKED ISSUE #3255: summarise shouldn't force base R functions like sum and mean
The `summarise` function forces the use of base R functions, instead of the expected behaviour that it would use loaded ('active') functions from the own environment.

```r
# (from another package:)
#' @inherit base::mean
#' @export
mean <- function(x, ..., na.rm = TRUE) {
  base::mean(x, ..., na.rm = na.rm)
}

library(dplyr)
df <- tibble(group = LETTERS[1:10], 
             score = c(runif(3, 0, 1), 
                       NA,
                       runif(6, 0, 1)))

mean(df$score) # now na.rm = TRUE by default
# [1] 0.5479927

df %>% summarise(m = mean(score))
# # A tibble: 1 x 1
#       m
#   <dbl>
# 1    NA

> df %>% summarise(m = mean(score, na.rm = TRUE))
# # A tibble: 1 x 1
#       m
#   <dbl>
# 1    0.5479927
```
Why not use functions from ones own environment? This `mean` function comes from another package, which I deliberately loaded, so I would expect it to work in all my code. But apparently, the `summarise` function is an exception. I think you should not force the use of base R functions like `base::mean` in dplyr.

You might think that package developers just shouldn't overwrite functions like `mean`, but even dplyr does this (like `intersect`, `setdiff`, `setequal`, `union`). That too gives other results than when only base R was loaded, just like you would expect. Hence, you loaded dplyr. Now I ask you to let `summarise` use the latest loaded functions, not just base R. Another example: if I would write a function to overwrite `sum` to solve #3189, it wouldn't be used by `summarise` at default, even when my package was loaded after dplyr.

### PR DESCRIPTION
**Pull request by romainfrancois**
- hybrid functions can now be masked by regular R functions to turn hybrid evaluation off (#3255). 

--- 

closes #3255 

@lionel- @krlmlr this follows the discussion in #3255. Some of the tests in `test-hybrid.R` had to change because they were based on the previous behavior. 

```
test_that("nth handlers support quosured symbols", {
  first <- last <- nth <- bad_hybrid_handler

  expect_identical(
    pull(summarise(mtcars, first(!!quo(cyl)))),
    dplyr::first(mtcars$cyl)
  )
  expect_identical(
    pull(summarise(mtcars, last(!!quo(cyl)))),
    dplyr::last(mtcars$cyl)
  )
  expect_identical(
    pull(summarise(mtcars, nth(!!quo(cyl), 2))),
    dplyr::nth(mtcars$cyl, 2)
  )
})

```

is now: 

```
test_that("nth handlers support quosured symbols", {
  expect_identical(
    pull(summarise(mtcars, first(!!quo(cyl)))),
    dplyr::first(mtcars$cyl)
  )
  expect_identical(
    pull(summarise(mtcars, last(!!quo(cyl)))),
    dplyr::last(mtcars$cyl)
  )
  expect_identical(
    pull(summarise(mtcars, nth(!!quo(cyl), 2))),
    dplyr::nth(mtcars$cyl, 2)
  )
})

```

otherwise, by design this calls the redefined functions `first`, `last` ...

Maybe we would need to revisit so that hybridness is tested

### CODE REVIEW

File: tests/testthat/test-hybrid.R [hunk 1]
```diff
@@ -1024,8 +1024,6 @@ test_that("constant folding and argument matching in hybrid evaluator (#2299)",
 })
 
 test_that("simple handlers supports quosured symbols", {
-  mean <- sum <- var <- sd <- bad_hybrid_handler
```
Comment (krlmlr) [inline tests/testthat/test-hybrid.R]: Would you mind adapting these tests to use `check_hybrid_result()` and friends? This testing API is under review (#2880), but for now I think it's the best we have for explicitly checking if hybrid evaluation kicks in.

Also, a second set of tests that checks that a local version of `mean()` etc. is used would be great.

File: tests/testthat/test-hybrid.R [hunk 2]
```diff
@@ -1130,6 +1117,88 @@ test_that("nth handlers support quosured symbols", {
 })
 
 test_that("top_n() is hybridised (#2822)", {
-  min_rank <- bad_hybrid_handler
   expect_error(top_n(mtcars, 1, cyl), NA)
 })
+
+test_that( "hybrid evaluation can be disabled locally (#3255)", {
+  tbl <- data.frame( x = 1:10)
```
Comment (krlmlr) [inline tests/testthat/test-hybrid.R]: Can you please use consistent coding style?

File: inst/include/dplyr/HybridHandler.h [hunk 3]
```diff
@@ -4,8 +4,24 @@
 namespace dplyr {
 class ILazySubsets;
 class Result;
-}
 
-typedef dplyr::Result* (*HybridHandler)(SEXP, const dplyr::ILazySubsets&, int);
+typedef dplyr::Result* (*HybridHandlerFun)(SEXP, const dplyr::ILazySubsets&, int);
```
Comment (krlmlr) [inline inst/include/dplyr/HybridHandler.h]: This now seems tied to `struct HybridHandler`, maybe use a local typedef?

File: src/hybrid_debug.cpp [hunk 4]
```diff
@@ -88,6 +88,8 @@ Result* verify_not_hybrid_prototype(SEXP call, const ILazySubsets&, int nargs) {
 }
 
 void install_debug_handlers(HybridHandlerMap& handlers) {
-  handlers[ Rf_install("verify_hybrid") ] = verify_hybrid_prototype;
-  handlers[ Rf_install("verify_not_hybrid") ] = verify_not_hybrid_prototype;
+  Environment ns_dplyr = Environment::namespace_env("dplyr") ;
```
Comment (krlmlr) [inline src/hybrid_debug.cpp]: Do we still need `ns_dplyr` here?

File: src/hybrid_debug.cpp [hunk 5]
```diff
@@ -88,6 +88,8 @@ Result* verify_not_hybrid_prototype(SEXP call, const ILazySubsets&, int nargs) {
...
+  // these are handled differently than the real hybrids
```
Comment (krlmlr) [inline src/hybrid_debug.cpp]: What's the consequence? Can you please expand the comment?

File: src/hybrid.cpp [hunk 6]
```diff
@@ -171,9 +173,16 @@ Result* get_handler(SEXP call, const ILazySubsets& subsets, const Environment& e
       return 0;
     }
 
+    // no hybrid evaluation if the symbol evaluates to something else than
+    // is expected. This would happen if e.g. the mean function has been shadowed
+    // mutate( x = mean(x) )
+    // if `mean` evaluates to something other than `base::mean` then no hybrid.
+    RObject fun = Rf_findFun(fun_symbol, env) ;
```
Comment (krlmlr) [inline src/hybrid.cpp]: Can a method in `HybridHandler` take care of this check?

File: src/hybrid.cpp [hunk 7]
```diff
@@ -150,12 +150,14 @@ Result* get_handler(SEXP call, const ILazySubsets& subsets, const Environment& e
     HybridHandlerMap& handlers = get_handlers();
 
     // interpret dplyr::fun() as fun(). #3309
+    bool check = true ;
```
Comment (krlmlr) [inline src/hybrid.cpp]: Can you please use a more explicit name?

File: tests/testthat/test-hybrid.R (hunk 1 above)
Comment (romainfrancois) [reply tests/testthat/test-hybrid.R]: I'm not sure what `check_hybrid_result` does. maybe I'll leave that one to you.  

I have added these second set of tests.

File: src/hybrid_debug.cpp (hunk 4 above)
Comment (romainfrancois) [reply src/hybrid_debug.cpp]: yes, see below

File: src/hybrid_debug.cpp (hunk 5 above)
Comment (romainfrancois) [reply src/hybrid_debug.cpp]: I reverted actually, they do the same as other hybrids.

File: src/hybrid.cpp (hunk 7 above)
Comment (romainfrancois) [reply src/hybrid.cpp]: I've used `check_hybrid_reference` now and added a comment

File: src/hybrid.cpp (hunk 6 above)
Comment (romainfrancois) [reply src/hybrid.cpp]: I like it here, it would otherwise mean passing things into HybridHandler which would then be more complicated.

**Comment by romainfrancois** [comment]
So I think that's ready @krlmlr

**Comment by krlmlr** [comment]
@lionel-: Would you like to review too?

**Comment by romainfrancois** [comment]
We could perhaps log the information somewhere in `get_handler`

**Comment by krlmlr** [comment]
We still have `verify_hybrid()`.?

**Comment by romainfrancois** [comment]
yes

**Comment by krlmlr** [comment]
Yeah, #2880 is pretty high on my list of prios ;-) I'll also work on documenting it better.

**Comment by krlmlr** [comment]
Thanks!

**Pull request merged**